File: producer_with_dsl_restdocs/src/test/java/com/example/intoxication/BeerIntoxicationBase.java
package com.example.intoxication;


import io.restassured.module.mockmvc.RestAssuredMockMvc;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.TestInfo;
import org.junit.jupiter.api.extension.ExtendWith;


import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.restdocs.RestDocumentationContextProvider;
import org.springframework.restdocs.RestDocumentationExtension;

import org.springframework.test.web.servlet.setup.MockMvcBuilders;
import org.springframework.web.context.WebApplicationContext;

import static com.example.intoxication.DrunkLevel.DRUNK;
import static com.example.intoxication.DrunkLevel.SOBER;
import static com.example.intoxication.DrunkLevel.TIPSY;
import static com.example.intoxication.DrunkLevel.WASTED;
import static org.springframework.restdocs.mockmvc.MockMvcRestDocumentation.document;
import static org.springframework.restdocs.mockmvc.MockMvcRestDocumentation.documentationConfiguration;


/**
 * Tests for the scenario based stub
 */
@SpringBootTest(classes = BeerIntoxicationBase.Config.class)
@ExtendWith(RestDocumentationExtension.class)

public abstract class BeerIntoxicationBase {

	@Autowired WebApplicationContext context;

	@BeforeEach
	public void setup(RestDocumentationContextProvider provider, TestInfo testInfo) {
		RestAssuredMockMvc.mockMvc(MockMvcBuilders.webAppContextSetup(this.context)
				.apply(documentationConfiguration(provider))
				.alwaysDo(document(getClass().getSimpleName() + "_" + testInfo.getDisplayName()))
				.build());
	}

	@Configuration
	@EnableAutoConfiguration
	static class Config {

		@Bean BeerServingController controller() {
			return new BeerServingController(responseProvider());
		}

		@Bean ResponseProvider responseProvider() {
			return new MockResponseProvider();
		}
	}

	
	static class MockResponseProvider implements ResponseProvider {

		private DrunkLevel previous = SOBER;
		private DrunkLevel current = SOBER;

		@Override public Response thereYouGo(Customer personToCheck) {
			
			if ("marcin".equals(personToCheck.name)) {
				 switch (this.current) {
				 case SOBER:
					 this.current = TIPSY;
					 this.previous = SOBER;
					 break;
				 case TIPSY:
					 this.current = DRUNK;
					 this.previous = TIPSY;
					 break;
				 case DRUNK:
					 this.current = WASTED;
					 this.previous = DRUNK;
					 break;
				 case WASTED:
					 throw new UnsupportedOperationException("You can't handle it");
				 }
			}
			
			return new Response(this.previous, this.current);
		}
	}
	
}

File: producer_kotlin/src/test/kotlin/com/example/intoxication/BeerIntoxicationBase.kt
package com.example.intoxication

import com.example.intoxication.DrunkLevel.*
import io.restassured.module.mockmvc.RestAssuredMockMvc
import org.junit.jupiter.api.BeforeEach
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.EnableAutoConfiguration
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.web.context.WebApplicationContext

/**
 * Tests for the scenario based stub
 */
 // remove::start[]
@SpringBootTest(classes = arrayOf(BeerIntoxicationBase.Config::class))
// remove::end[]
abstract class BeerIntoxicationBase {

    // remove::start[]
    @Autowired
    internal var webApplicationContext: WebApplicationContext? = null

    @BeforeEach
    fun setup() {
        RestAssuredMockMvc.webAppContextSetup(this.webApplicationContext)
    }

    @Configuration
    @EnableAutoConfiguration
    open class Config {

        @Bean
        open fun controller(): BeerServingController {
            return BeerServingController(responseProvider())
        }

        @Bean
        open fun responseProvider(): ResponseProvider {
            return MockResponseProvider()
        }
    }

    internal class MockResponseProvider : ResponseProvider {

        private var previous = SOBER
        private var current = SOBER

        override fun thereYouGo(personToCheck: Customer): Response {
            if ("marcin" == personToCheck.name) {
                when (this.current) {
                    SOBER -> {
                        this.current = TIPSY
                        this.previous = SOBER
                    }
                    TIPSY -> {
                        this.current = DRUNK
                        this.previous = TIPSY
                    }
                    DRUNK -> {
                        this.current = WASTED
                        this.previous = DRUNK
                    }
                    WASTED -> throw UnsupportedOperationException("You can't handle it")
                }
            }
            return Response(this.previous, this.current)
        }
    }
    // remove::end[]
}
